oss_src/unity/lib/unity_sarray_builder.cpp

```diff
@@ -0,0 +1,114 @@
+/**
+ * Copyright (C) 2015 Dato, Inc.
+ * All rights reserved.
+ *
+ * This software may be modified and distributed under the terms
+ * of the BSD license. See the LICENSE file for details.
+ */
+
+#include <unity/lib/unity_sarray_builder.hpp>
+#include <unity/lib/unity_sarray.hpp>
+
+namespace graphlab {
+
+void unity_sarray_builder::init(size_t num_segments, size_t history_size, flex_type_enum dtype) {
+  if(m_inited)
+    log_and_throw("This sarray_builder has already been initialized!");
+
+  m_sarray = std::make_shared<sarray<flexible_type>>();
+  m_sarray->open_for_write(num_segments);
+  m_out_iters.resize(num_segments);
+  for(size_t i = 0; i < num_segments; ++i) {
+    m_out_iters[i] = m_sarray->get_output_iterator(i);
+  }
+  m_history = std::make_shared<boost::circular_buffer<flexible_type>>(history_size);
+  m_given_dtype = dtype;
+  if(dtype != flex_type_enum::UNDEFINED)
+    m_ary_type = m_given_dtype;
+
+  m_inited = true;
+}
+
+void unity_sarray_builder::append(const flexible_type &val, size_t segment) {
+  if(!m_inited)
+    log_and_throw("Must call 'init' first!");
+
+  if(m_closed)
+    log_and_throw("Cannot append values when closed.");
+
+  if(segment >= m_out_iters.size()) {
+    log_and_throw("Invalid segment number!");
+  }
+
+  m_history->push_back(val);
+  auto in_type = val.get_type();
+  // Only care about types if a dtype was not provided
+  if(m_given_dtype == flex_type_enum::UNDEFINED &&
+      in_type != flex_type_enum::UNDEFINED) {
+    auto ins_ret = m_types_inserted.insert(in_type);
```

Comment:
If only 1 type is allowed, we don't need m_types_inserted.
Just need to compare in_type vs m_ary_type.